File: semester_3/hw2/Task_1/Task_1/Network.cs
namespace Task_1
{
    using System;
    using System.IO;
    using System.Linq;

    /// <summary>
    /// Local network.
    /// </summary>
    public class Network : INetwork
    {
        /// <summary>
        /// Field for generating random values.
        /// </summary>
        private Random random = new Random();

        /// <summary>
        /// Initializes a new instance of the <see cref="Network"/> class.
        /// </summary>
        /// <param name="filepath">Path of file.</param>
        public Network(string filepath) => this.LoadInformationFromFile(filepath);

        /// <summary>
        /// Gets the list of computers.
        /// </summary>
        public IComputer[] Computers { get; private set; }

        /// <summary>
        /// Gets the graph of computers dependencies.
        /// </summary>
        public bool[,] Graph { get; private set; }

        /// <summary>
        /// Getting number of infected computers.
        /// </summary>
        /// <returns>Number of infected computers.</returns>
        public int NumberOfInfectedComputers() => this.Computers.Count(computer => computer.IsInfected);

        /// <summary>
        /// Making move.
        /// </summary>
        public void MakeMove()
        {
            if (this.AreAllNotInfected())
            {
                this.TryToInfectRandomComputer();
                return;
            }

            if (this.AreAllInfected())
            {
                return;
            }

            // New bool array which shows if computer with index i is infected.
            var infectedComputers = new bool[this.Computers.Length];

            for (var i = 0; i < this.Computers.Length; ++i)
            {
                if (this.Computers[i].IsInfected)
                {
                    infectedComputers[i] = true;
                }
            }

            // Trying to infect computers.
            for (var i = 0; i < this.Computers.Length; ++i)
            {
                if (infectedComputers[i])
                {
                    for (var j = 0; j < this.Graph.GetLongLength(0); ++j)
                    {
                        if (this.Graph[i, j])
                        {
                            this.Computers[j].TryToInfect();
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Printing information about all computers in the network.
        /// </summary>
        public void PrintInformation()
        {
            for (var i = 0; i < this.Computers.Length; ++i)
            {
                Console.WriteLine($"Computer {i + 1} on {this.Computers[i].OS} infection: {this.Computers[i].IsInfected}");
            }

            Console.WriteLine();
        }

        /// <summary>
        /// Loading some information from file: number of computers, OS, dependencies.
        /// </summary>
        /// <param name="filepath">Path of file.</param>
        private void LoadInformationFromFile(string filepath)
        {
            try
            {
                // Open the text file using a stream reader.
                using (var sr = new StreamReader(filepath))
                {
                    int size = int.Parse(sr.ReadLine());

                    this.Computers = new IComputer[size];
                    this.Graph = new bool[size, size];
            
                    var text = sr.ReadLine();
                    string[] bits = text.Split(' ');

                    for (var i = 0; i < size; ++i)
                    {
                        this.Computers[i] = new Computer((OS)int.Parse(bits[i]), random);
                    }

                    for (var i = 0; i < size; ++i)
                    {
                        text = sr.ReadLine();
                        bits = text.Split(' ');
                        for (var j = 0; j < size; ++j)
                        {
                            this.Graph[i, j] = Convert.ToBoolean(int.Parse(bits[j]));
                        }
                    }
                }
            }
            catch (IOException e)
            {
                Console.WriteLine("The file could not be read:");
                Console.WriteLine(e.Message);
            }
        }

        /// <summary>
        /// Checking all computers on infecting.
        /// </summary>
        /// <returns>True – all infected, false – otherwise.</returns>
        private bool AreAllInfected() => this.NumberOfInfectedComputers() == this.Computers.Length;

        /// <summary>
        /// Checking all computers on infection.
        /// </summary>
        /// <returns>True – all not infected, false – otherwise.</returns>
        private bool AreAllNotInfected() => this.NumberOfInfectedComputers() == 0;

        /// <summary>
        /// Trying to infect random computer.
        /// </summary>
        private void TryToInfectRandomComputer()
        {
            var index = this.random.Next(0, this.Computers.Length);
            this.Computers[index].TryToInfect();
        }
    }
}

File: semester_4/hw6/Task_1/Task_1/Network.fs
module Network
open Computer
open System
open System.IO

/// <summary>
/// Interface for network class.
/// </summary>
type INetwork = 

    /// <summary>
    /// Contains array of computers.
    /// </summary>
    abstract Computers : IComputer[]

    /// <summary>
    /// Contains adjacency matrix.
    /// </summary>
    abstract Graph : bool[][]

    /// <summary>
    /// Makes move: tries to infect computer.
    /// </summary>
    abstract MakeMove : unit -> unit

    /// <summary>
    /// Print detailed information: computer ID and whether it is infected.
    /// </summary>
    abstract PrintInformation : unit -> unit

    /// <summary>
    /// Gets number of infected computers.
    /// </summary>
    abstract NumberOfInfectedComputers : unit -> int

/// <summary>
/// Network class.
/// Contains all methods for performing basic network functions.
/// </summary>
type Network(computers : IComputer[], graph : bool[][], random : Random) = 
    
    let numberOfInfectedComputers () = 
        computers |> Array.filter (fun computer -> computer.IsInfected) |> Array.length
    let areAllInfected () =
        Array.length computers = numberOfInfectedComputers ()
    let areAllNotInfected () = 
        numberOfInfectedComputers () = 0
    let tryToInfectRandomComputer () =
        let index = random.Next(0, Array.length computers)
        computers.[index].TryToInfect ()

    /// <summary>
    /// Loads essential information from file: number of computers, operation systems, adjacency matrix.
    /// </summary>
    /// <param name="filePath">File path.</param>
    new (filePath : string) = 
        let random = new Random()
        let lines = File.ReadAllLines filePath
        let count = lines.[0].Split() |> Array.item 0 |> int
        let os = lines.[1].Split() |> Array.map (int >> enum)
        let computers = [| for index in 0 .. count - 1 -> new Computer(os.[index], random) :> IComputer |]
        let graph = 
            [| 
                for line in (lines |> Array.toSeq |> Seq.skip 2) do
                yield line.Split() |> Array.map (fun x -> if x = "1" then true else false)
            |]
        Network(computers, graph, random)
    
    interface INetwork with
        member this.Computers = computers
        member this.Graph = graph
        member this.MakeMove () =
            match areAllNotInfected (), areAllInfected () with
            | true, _ -> tryToInfectRandomComputer ()
            | _, false -> let infectRow row = graph.[row] |> Array.iteri (fun column state -> if (graph.[row].[column]) then computers.[column].TryToInfect ())
                          computers |> Array.iteri (fun index computer -> if computer.IsInfected then infectRow index)
            | _, _ -> ()
        member this.PrintInformation () =
            computers |> Array.iteri (fun index computer -> printfn "Computer %i on %O infection: %b" index computer.OS computer.IsInfected)
        member this.NumberOfInfectedComputers () = numberOfInfectedComputers ()